apps/src/music/blockly/MusicBlocklyWorkspace.js

```diff
@@ -116,11 +118,66 @@ export default class MusicBlocklyWorkspace {
 
     const events = {};
 
-    this.workspace.getTopBlocks().forEach(block => {
-      if (block.type === BlockTypes.WHEN_RUN) {
+    const topBlocks = this.workspace.getTopBlocks();
+
+    if (getBlockMode() === BlockMode.SIMPLE2) {
+      // If there's no when_run block, then we'll generate
+      // some custom code that calls all the functions
+      // together, simulating tracks mode.
+
+      if (
+        !topBlocks.some(block => block.type === BlockTypes.WHEN_RUN_SIMPLE2)
+      ) {
         events.whenRunButton = {
-          code: Blockly.JavaScript.blockToCode(block)
+          code: `
+              ProgramSequencer.init();
+              ProgramSequencer.playTogether();
+            `
         };
+
+        topBlocks.forEach(functionBlock => {
+          if (functionBlock.type === 'procedures_defnoreturn') {
+            events.whenRunButton.code += `${functionBlock.getFieldValue(
+              'NAME'
+            )}();
+            `;
+          }
+        });
+      }
+    }
+
+    topBlocks.forEach(block => {
+      if (getBlockMode() !== BlockMode.SIMPLE2) {
+        if (block.type === BlockTypes.WHEN_RUN) {
+          events.whenRunButton = {
+            code: Blockly.JavaScript.blockToCode(block)
+          };
+        }
+      } else {
+        if (block.type === BlockTypes.WHEN_RUN_SIMPLE2) {
+          if (!events.whenRunButton) {
+            events.whenRunButton = {code: ''};
+          }
+          events.whenRunButton.code += Blockly.JavaScript.blockToCode(block);
+        }
+
+        if (block.type === 'procedures_defnoreturn') {
+          if (!events.whenRunButton) {
+            events.whenRunButton = {code: ''};
+          }
+
+          const functionCode = Blockly.JavaScript.blockToCode(
+            block.getChildren()[0]
+          );
+          events.whenRunButton.code += `function ${block.getFieldValue(
+            'NAME'
+          )}() {
+              ProgramSequencer.playSequential();
+              ${functionCode}
+              ProgramSequencer.endSequential();
+            }
+            `;
```

Comment:
I can't think of a way around this, but having these hard-coded strings feels fragile. Is there another way this can be approached?

Reply:
I know what you mean.  When there was more of this code to be put in the interpreter I exported it as a string from the `simple2.js` blocks file, but it also didn't feel like quite the right approach.  This `executeSong` function should get a fair bit smaller and simpler once we consolidate some of these model explorations in the next couple weeks, so we could revisit then.

Reply:
Good point, and for now I've at least moved the this type of code, which feels like a kind of special-case "generator", into a helper class in `blocks/simple2.js`.  We might be able to figure out a better pattern moving forward, but it at least keeps all the hand-crafted generator code in one source file that is model-specific.

Reply:
Update made in the follow-up PR.